src/main/java/ch/njol/skript/expressions/ExprDifference.java

```diff
@@ -33,123 +28,159 @@
+		ClassInfo<?> classInfo = Classes.getSuperClassInfo(Utils.getSuperType(firstReturnType, secondReturnType));
+
+		boolean fail = false;
+
+		if (classInfo.getC() == Object.class && (firstReturnType != Object.class || secondReturnType != Object.class)) {
+			// These two types may not be compatible
+			if (firstReturnType != Object.class && secondReturnType != Object.class) {
+				// Let's see if we can find something equivalent using converters
+				Expression<?> secondConverted = second.getConvertedExpression(firstReturnType);
+				if (secondConverted == null) {
+					Expression<?> firstConverted = first.getConvertedExpression(secondReturnType);
+					if (firstConverted == null) {
+						// Not even converters could save them :(
+						// Because we have two known types, we need to fail
+						fail = true;
+					} else {
+						first = firstConverted;
+					}
+				} else {
+					second = secondConverted;
+				}
+			} else { // Avoid converting an actual type into Object.class
+				Expression<?> converted;
+				if (firstReturnType == Object.class) {
+					converted = first.getConvertedExpression(secondReturnType);
+					if (converted != null) {
+						first = converted;
+					}
+				} else {
+					converted = second.getConvertedExpression(firstReturnType);
+					if (converted != null) {
+						second = converted;
+					}
+				}
+
+				if (converted == null) { // It's unlikely that these two can be compared
+					fail = true;
+				}
 			}
-			if (first instanceof Variable) {
-				first = first.getConvertedExpression(second.getReturnType());
-			} else if (second instanceof Variable) {
-				second = second.getConvertedExpression(first.getReturnType());
+
+			if (!fail) { // Now that we've converted, let's aim for a better class info
+				classInfo = Classes.getSuperClassInfo(Utils.getSuperType(first.getReturnType(), second.getReturnType()));
 			}
-			assert first != null && second != null;
-			ci = Classes.getSuperClassInfo(Utils.getSuperType(first.getReturnType(), second.getReturnType()));
-		}
-		assert ci != null;
-		if (!ci.getC().equals(Object.class) && ci.getMath() == null) {
-			Skript.error("Can't get the difference of " + CondCompare.f(first) + " and " + CondCompare.f(second), ErrorQuality.SEMANTIC_ERROR);
-			return false;
+
 		}
-		if (ci.getC().equals(Object.class)) {
+
+		if (classInfo.getC() == Object.class) {
 			// Initialize less stuff, basically
 			relativeType = Object.class; // Relative math type would be null which the parser doesn't like
+		} else if (classInfo.getMath() == null || classInfo.getMathRelativeType() == null) {
+			fail = true;
 		} else {
-			math = ci.getMath();
-			relativeType = ci.getMathRelativeType();
+			math = classInfo.getMath();
+			relativeType = classInfo.getMathRelativeType();
 		}
+
+		this.first = first;
+		this.second = second;
+
+		if (fail) {
+			Skript.error("Can't get the difference of " + CondCompare.f(first) + " and " + CondCompare.f(second), ErrorQuality.SEMANTIC_ERROR);
```

Comment:
It's `SEMANTIC_ERROR` by default
```suggestion
			Skript.error("Can't get the difference of " + CondCompare.f(first) + " and " + CondCompare.f(second));
```